### PR TITLE
Custom field form reform

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
This overhauls the admin form for creating/updating custom fields, with improved validation, easier-to-select defaults, and more flexibility about changing the widget type.

Before
----------------------------------------
- Changing the widget type was a separate form

After
----------------------------------------
- Changing the widget type integrated into the main form, with improved validation and feedback about option mismatches.

Comments
----------------------------------------
This PR will need some manual testing. Some workflows to test:

1. Create a multi-select or checkbox custom field. Populate it with some data for a few contacts, and for at least one contact select multiple options. Go back and change the field type to single-select. You should get a warning about the number of records that will lose data because they contain multiple values.
2. Create a text input field. Populate a few contacts with data. Go back and try to change it to a field type with an option list. You'll get an error unless the options you entered match the existing data exactly.
3. Try creating a Country, State, Yes/No, or Date field - setting the default value is now more user-friendly.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@highfalutin you've created a monster...

**Comment by highfalutin** [comment]
@eileenmcnaughton sorry? 😂

**Comment by colemanw** [comment]
test this please

**Comment by colemanw** [comment]
I know @eileenmcnaughton this is a big diff. (would be smaller but it includes #18356). It's not the sort of thing to review line-by-line as it's overhauling a quickform with a lot of messy code. It just needs to be run.

**Comment by eileenmcnaughton** [comment]
I'm hoping @highfalutin will do UI testing as opening this Pandora's box...

**Comment by eileenmcnaughton** [comment]
Just noting that I agree ui testing is what is required to get this merged. I don't have concerns about the code or test cover & the impact can be judged by testing the form used to manage custom fields & in particular changing between field types

**Comment by kcristiano** [comment]
@colemanw Just to be sure before I wade into testing - we do expect data loss on item 1 when converting, am I correct?  

I am uneasy about that to be frank.  Do we have a gitlab discussion on this?

**Comment by colemanw** [comment]
@kcristiano This PR is mostly just a UI change. You may want to familiarize yourself with the previous UI, in which it was also possible to change a Multi-Select to a Single-Select (by clicking on "Change input field type" while editing a custom field). Yes, doing so would cause data-loss, and still does. However I think you'll find the new UI in this PR does a better job of warning you about just how much data will be lost before allowing you to proceed.

**Comment by kcristiano** [comment]
Thanks @colemanw That makes sense, I'll give this a test.

**Comment by kcristiano** [comment]
@colemanw 1 and 3 work exactly as expected.

> Create a text input field. Populate a few contacts with data. Go back and try to change it to a field type with an option list. You'll get an error unless the options you entered match the existing data exactly.

I created a text input field and I cannot change the field type at all.   Am I doing something wrong, perhpas a screen shot of what is expected here would help.

**Comment by colemanw** [comment]
**1. Create text input field**
image

**2. Populate some data (which will ultimately match your option values)**
image

**3. Change the field to a select, matching the option values to the data you entered:**
image

If your options *don't* match existing data you should get an error.

**Comment by colemanw** [comment]
@kcristiano I don't think it was possible to change text field type prior to this PR, so if you're trying to do a "before" comparison then it won't work. This PR makes it possible and also adds validation to make it safe.

**Comment by kcristiano** [comment]
I am doing the 'after PR' work.

My setups don't have that option (single line input) 5.31 + PR:

image

If I set it up and try and change  There is no option to change:

image

**Comment by kcristiano** [comment]
@colemanw Never mind.   The screens now show up properly - not sure what happened before, but a restart of my local dev has it appearing now:

image


r-run shows this to work as expected.

Would love more UI review, but it looks good to me.

File: CRM/Custom/Form/Field.php [hunk 1]
```diff
@@ -947,24 +844,19 @@ public function postProcess() {
...
+    if ($params['html_type'] === 'Select') {
       $params['serialize'] = $params['serialize'] ? CRM_Core_DAO::SERIALIZE_SEPARATOR_BOOKEND : 'null';
     }
     else {
       $params['serialize'] = $params['html_type'] == 'CheckBox' ? CRM_Core_DAO::SERIALIZE_SEPARATOR_BOOKEND : 'null';
     }
 
     $filter = 'null';
-    if ($dataTypeKey == 11 && !empty($params['filter_selected'])) {
+    if ($params['date_type'] == 'ContactReference' && !empty($params['filter_selected'])) {
```
Comment (demeritcowboy) [inline CRM/Custom/Form/Field.php:855]: data_type

File: CRM/Core/DAO/AllCoreTables.php [hunk 2]
```diff
@@ -280,7 +280,7 @@ public static function getClasses() {
    * Get the classname for the table.
    *
    * @param string $tableName
-   * @return string
+   * @return CRM_Core_DAO|NULL
    */
```
Comment (demeritcowboy) [inline CRM/Core/DAO/AllCoreTables.php:284]: The return *is* a string?

**Comment by eileenmcnaughton** [comment]
@colemanw please self-merge this once you have addressed @demeritcowboy's comments - @kcristiano seems mostly adequate (bar missing the date_type - but we got that spotted now :-))

File: CRM/Custom/Form/Field.php (hunk 1 above)
Comment (colemanw) [reply CRM/Custom/Form/Field.php]: Good catch!

File: CRM/Core/DAO/AllCoreTables.php (hunk 2 above)
Comment (colemanw) [reply CRM/Core/DAO/AllCoreTables.php:284]: For convenience of IDE auto-completion, we kinda hack the `@return` annotationwhen it's returning a string which is also a class name. In php strings can be used interchangeably with classes, go figure.

**Comment by agileware-justin** [comment]
https://lab.civicrm.org/dev/core/-/issues/2214

**Pull request merged**